helper/cert/source_gen.go

```diff
@@ -143,9 +150,19 @@ func (s *GenSource) Certificate(ctx context.Context, last *Bundle) (Bundle, erro
 		result.Key = []byte(key)
 
 		if s.LeaderElector != nil {
-			if err := s.updateSecret(ctx, result); err != nil {
-				return result, fmt.Errorf("failed to update Secret: %s", err)
+			// Cancel and wait for the previous retryUpdateSecret to quit
+			if s.UpdateCancel != nil && s.StopUpdate != nil {
+				s.Log.Trace("cancelling cert secret update context")
+				s.UpdateCancel()
+				<-s.StopUpdate
+			}
+			// Start a new update goroutine
+			var updateCtx context.Context
+			updateCtx, s.UpdateCancel = context.WithCancel(ctx)
+			if s.StopUpdate == nil {
+				s.StopUpdate = make(chan struct{})
 			}
+			go s.retryUpdateSecret(updateCtx, result)
```

Comment:
AFAIU, this is run in a goroutine to allow the function to return the certificate before the k8s secret is updated, in case that takes a large number of retries? And that seems to add a fair bit of complexity and statefulness to the control flow, because now you have to synchronise across multiple calls to `Certificate`.

I definitely could be missing some concerns or requirements, but is that complexity necessary? It seems like it would be ok to synchronously update the k8s secret before returning, and then you could remove a bunch of the syncing required.

Reply:
I agree with this sentiment, and it makes the guarantee of `result` a bit weak on whoever calls `Certificate()` since the update on the k8s is not guaranteed to have occurred (or can just flat out error) before the caller consumes the value.

Reply:
Yeah, I think those are good points, and `Certificate()` is already being called from another goroutine anyway. Changed in 51f676243f146b93c2836e9905f09ff69e15a00b, and set the max time for retrying to be 30 minutes.